### ISSUE TITLE
Missing labels in batches when using small-width objects

### ISSUE THREAD

**Issue by r-romito** [ultralytics/ultralytics#1960]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Hi, I'm trying to train a dataset where objects are generally 1/2px wide and height may vary.

This is my current command line to start training:
`yolo train model=yolov8m.pt data=D:\yolo\train\data.yaml epochs=5 batch=5 scale=0 hsv_v=0 hsv_s=0 hsv_h=0 mosaic=0 translate=0 perspective=0 plots=True verbose=True
`
The problem is that if I tag my objects pixels-perfect (1px wide) I get untagged objects in the train_batch*.jpg files (no red bars), is that a bug related to the batches images generation or do I have to change some other parameters in order to train this kind of dataset?

I made some other experiments inflating the bounding boxes of my objects with 3px wide (1px margin), 5x wide (2px margin) and 7px wide (3px margin) and from what I see only starting from 7px wide I have a 100% of tags in train_batch*.jpg images while **the validation batches are always correctly tagged even when scaled (using 1,2 or 3 pixels margins)**

Here it is the result:

1. 1px margin (objects width = 3px total -> missing a lot of white lines):
![train_batch0](https://user-images.githubusercontent.com/105412400/231156128-c34c9cf2-e2ec-41fa-95ab-1fbb1ab4fb93.jpg)
![train_batch1](https://user-images.githubusercontent.com/105412400/231156135-9861cced-949a-47c9-9dc4-26cb000b0f45.jpg)
2. 2px margin (objects width = 5px total -> still missing some white lines)
![train_batch0](https://user-images.githubusercontent.com/105412400/231156943-c5f6c09e-2a40-4ab3-8e92-e3436ce1c724.jpg)
![train_batch1](https://user-images.githubusercontent.com/105412400/231156948-1ddb806c-7ea9-4ad3-8acb-ab629162e657.jpg)
3. 3px margin (objects width = 7px total -> no missing objects)
![train_batch0](https://user-images.githubusercontent.com/105412400/231157090-7032c246-c137-451b-b3c8-0d86a8ea4954.jpg)
![train_batch1](https://user-images.githubusercontent.com/105412400/231157097-446e24d1-d896-4584-8287-18c45f111dd0.jpg)

Can someone help me in understanding if the issue is only related to the generation of batch images or if traning objects less than ~7px in width is not allowed/requires some parameters tuning? I would like to avoid adding such "large" margins.

### Environment

- YOLO: Ultralytics YOLOv8.0.73  Python-3.7.9 torch-1.13.1+cu117 CUDA:0 (NVIDIA T1200 Laptop GPU, 4096MiB)
- OS: Windows 10
- Python: 3.7.9

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@r-romito also check the candidate labels filtering function here, this will remove label that do not meet candidate criteria, which include a minimum height or width of 2 pixels. You can modify these settings to suit your custom use case:

https://github.com/ultralytics/ultralytics/blob/e5cb35edfc3bbc9d7d7db8a6042778a751f0e39e/ultralytics/yolo/data/augment.py#L381-L387

**Comment by r-romito** [comment]
> box_candidates

@glenn-jocher thank you, that really helps!
Some questions:

1. What about making "wh_thr=2" configurable (maybe a CLI argument would be great)? 
2. Is there any log message provided (I'm on YOLOv8.0.73) so that I can inspect if I have some "filtered" objects by box_candidates or any other filter function? I abolutely need to avoid having untagged objects in my dataset.
3. Can you confirm that box_candidates is not called for the generation of the validation batches images? 

Thanks again, I'm not so familiar with the YOLO source code to provide a PR in particular for #1 (hope to see it in future releases)

**Comment by glenn-jocher** [comment]
@r-romito 1. `find_unused=True` is a PyTorch parameter, and its role is to detect unused or duplicated tensors in computation graphs, which can be caused by data or code issues. It’s not related to the `wh_thr` issue, but we recommended it because it can help identify any issues in the loss calculation, which may impact training. 
2. By default, YOLOv8 does not log the “Filtered” warning message. You can override this behavior by updating your code to print a warning message when the filter function removes any boxes. You can modify the log function in `yolov5/utils/logging.py` to print these messages to the console. By default, YOLOv8 does not log to a file, but you can override this to write the logs out to a file. The logs are written to the console, and you can use a log aggregation tool like ELK stack, Graylog, or Fluentd to save these logs to a file if desired.

**Comment by r-romito** [comment]
@glenn-jocher thanks for the details, I'll have a look on them.

Last "request" on my side: EDIT: what about avoiding to filter totally any object in RandomPerspective augmentator (to avoid this issues) when the YOLOv8 parameter perspective (as per my command line) is 0? To me this means "disable it" and it should not impact the dataset in any way. What about?

Does it make sense to change this code:

```pre_transform = Compose([
        Mosaic(dataset, imgsz=imgsz, p=hyp.mosaic, border=[-imgsz // 2, -imgsz // 2]),
        CopyPaste(p=hyp.copy_paste),
        RandomPerspective(
            degrees=hyp.degrees,
            translate=hyp.translate,
            scale=hyp.scale,
            shear=hyp.shear,
            perspective=hyp.perspective,
            pre_transform=LetterBox(new_shape=(imgsz, imgsz)),
        )])
```
To something like:

```pre_transform = Compose([])
if hyp.mosaic > 0:
   pre_transform .append(Mosaic(dataset, imgsz=imgsz, p=hyp.mosaic, border=[-imgsz // 2, -imgsz // 2]))
if hyp.copy_paste > 0:
   pre_transform .append(CopyPaste(p=hyp.copy_paste))
if hyp.perspective > 0:
   pre_transform .append(RandomPerspective(
            degrees=hyp.degrees,
            translate=hyp.translate,
            scale=hyp.scale,
            shear=hyp.shear,
            perspective=hyp.perspective,
            pre_transform=LetterBox(new_shape=(imgsz, imgsz)),
        ))
```

Does it impact badly somewhere? If it's fine I can provide a PR of course but (again) I'm not very familiar with the base code and the above suggestion may not be applicable.

**Comment by r-romito** [comment]
@glenn-jocher great! I'll try to provide that.

**Comment by r-romito** [comment]
PR [opened](https://github.com/ultralytics/ultralytics/pull/2003)!

Before changes (missing labels of small objects):
![train_batch0](https://user-images.githubusercontent.com/105412400/231749356-87d3cf76-f033-4706-b079-f8bf1c54638f.jpg)
![train_batch1](https://user-images.githubusercontent.com/105412400/231749365-bf6699fe-fdc6-4dad-8a2a-863e43da8ea9.jpg)
![train_batch2](https://user-images.githubusercontent.com/105412400/231749367-0912c159-f512-4699-bc03-edc2f34d8d09.jpg)

After the changes (small objects are not filtered!):
![fixed_train_batch0](https://user-images.githubusercontent.com/105412400/231749531-66cfc1e9-56a9-4a6a-8537-4acfc759f81a.jpg)
![fixed_train_batch1](https://user-images.githubusercontent.com/105412400/231749537-52058c7c-8b7d-451b-bda5-b46967a7903c.jpg)
![fixed_train_batch2](https://user-images.githubusercontent.com/105412400/231749539-ca2554b0-8a83-4727-b75a-a3bfc1ebcf01.jpg)

Many thanks @glenn-jocher for pointing me in the right direction, I hope to see the PR integrated in the main branch and contribute again in the future to the project.

**Comment by glenn-jocher** [comment]
@r-romito great job on your PR! Your proposal to optimize the `RandomPerspective()` augmentation in YOLOv8 to allow for disabling it with the `perspective` parameter is a useful improvement that will help users easily turn off an augmentation technique that they don't need. The changes you've made to the `box_candidates` function also improve object detection and reduce the likelihood of small objects being filtered out in the training data. Allowing users to configure the `wh_thr` parameter is another important feature that increases the flexibility and customization of the YOLOv8 training pipeline. Overall, your contributions will make YOLOv8 more user-friendly and efficient for training object detection models, and I hope to see more of your contributions in the future.
